src/Twine/Guard.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+module Twine.Guard (
+    TerminationAction (..)
+  , TerminationHandler (..)
+  , guarded
+  , repeatedly
+  ) where
+
+import           Control.Monad.Catch
+import           Control.Monad.Trans.Either
+import           Control.Monad.Trans.Class
+
+import           P
+
+import           System.IO
+
+import           Twine.Snooze
+
+data TerminationAction =
+    Restart
+  | Die
+  deriving (Eq, Show)
+
+data TerminationHandler e =
+  TerminationHandler {
+      onExplosion :: SomeException -> IO TerminationAction
+    , onError :: e -> IO TerminationAction
+    , onGraceful :: IO TerminationAction
+    }
+
+-- | Run an action forever, using termination handler to notify of
```

Comment:
Those are the two situations I am trying to prevent from happening, so I really don't want to add them. The TerminationAction is to allow for tests to exit in a non-abrupt way. I was slightly loath to add it, because anything in production not using restart is almost certainly a bug. Open to other suggestions, but there needs to be a way to keep actions on a thread from ever hitting thread-finalizer and exiting, so adding something that encourages termination is something I really don't want to do, because it is the exact bug that I am trying to prevent that is in half a dozen of the code bases at the moment.